### DISCUSSION TITLE
Looking For OpenKH's KH2FM Hash-Pairs List For KH2MapStudio!

### DISCUSSION THREAD

**Discussion by Shinobz** [OpenKH/OpenKh#970] [Q&A] [answered]
Hello! 
I have some custom .IDX entries which I've successfully put into the KH2FM.iso as Player and Friend characters; "P_XEMNAS" and "F_XEMNAS" for example, both complete with their own .mdlx/.mset/.a.fm... 
I've got multiple playable characters set up like this, but now the problem is that I can't spawn them in the KH2MapStudio because they don't appear in the drop-down list. 

In CrazyCatz00's Patcher, the extract function works based on a list of hash-pairs found in "_Hashpairs.txt". By adding my new custom hashes to that document, his patcher is now able to successfully extract my new files. 

I'm assuming that OpenKH uses something similar to that list of hash-pairs for extraction, since when I extract my .iso with OpenKHModsManager, it doesn't extract my new extra files. 

I found a document in the "Resources" folder titled "KH2idx.txt"... This appears to be a document containing all of the paths/names of the files in the game for every region, however it contains no hash ID's. I tried to just add the path/name of my custom files to this document and extract again using OpenKHModsManager, but the result was still the same - that KH2MapStudio is not listing those files in the drop-down lists. 

I would really like to resolve this, as I just took the time to add a bunch of playable characters and party members to the .IDX, but now cannot spawn them how I need to. 

My end-goal here is to find the file which contains the hash-pair list, so that I can update that list, and see my files selectable/spawnable in KH2MapStudio.

It is worth noting that after adding these files to the "00objentry.bin" with OpenKH's ObjEntryEditor, they also show up in OpenKH's 03SystemEditor, where I see them perfectly selectable in the drop-down list.

I've attached some photos as well:
CrazyCatz00's "_Hashpairs.txt" document -
![crazycatz00's _Hashpairs txt document](https://github.com/OpenKH/OpenKh/assets/150029221/e79a2b27-7608-464c-9fcb-69630cce137a)
OpenKH's "KH2idx.txt" document -
![OpenKH's KH2idx txt document](https://github.com/OpenKH/OpenKh/assets/150029221/e0b98b12-9f4e-4639-8d86-d2f579a0f32b)

Thank you in advance for your time! 🙏

**Comment by Vladabdf** [comment]
> In CrazyCatz00's Patcher

Why would you use that instead of the OpenKH mod manager? It supports patching if that's really something you need (which you only should if you're modding a game to be played on retail hardware), as well as a dynamic file loader for mods. Both of these remove the need for the hash system entirely.

That KH2idx document is only used within the OpenKH suite as a point of reference for the hashes built into shared code for various IDX/IMG functions. It is never something you will never need to touch manually, as editing the IDX properly with the IDXIMG tool should generate a proper hash for you and insert it into your hash list. Again though, you shouldn't need to do that anyway. 

Legacy tooling like the older patcher system are only supported by OpenKH in the capacity of our mod manager being able to read those mods; no further support for the old methods will be added.

**Reply by Shinobz** [reply]
Thank you so much for your reply! 🙏

> Legacy tooling like the older patcher system are only supported by OpenKH in the capacity of our mod manager being able to read those mods; no further support for the old methods will be added.

So a couple of things for clarification regarding my mention of CrazyCatz00's Patcher... I do not require any support or assistance with that legacy tool, **I'm only here for support regarding the KH2MapStudio not showing my custom model entries in it's drop-down lists.** I only brought up CrazyCatz00's Patcher as reference to what I was talking about in terms of modifying a hash-pair list, and also to give information on what it is I've actually succeeded to do thus far. I've not only added the files to the .IDX with new hashes, but I've also successfully re-extracted the game with those hashes and verified the files are within the .iso; So I now have an extracted copy and an unextracted copy of KH2FM with my custom entries inside.  When I use the OpenKHModsManager Wizard for extraction, it completely skips over these new files which didn't originally exist in the .IDX.

> It supports patching if that's really something you need (which you only should if you're modding a game to be played on retail hardware)

Patching will be necessary for my mod in order to make all changes final, since I am making this mod with the intent of it being playable as a fully native ROM hack which can be played simply via .iso on as many devices as possible, including PS2, PS3, PS4, PS5, XBX, PC, Mac, Linux, & certain mobile devices. For this reason, I'm also using no cheat codes and no external tools which will need to run alongside it... However, it is worth noting that I'm not here for support with patching, that part is already figured out; **my only concern is getting the patched files (which didn't originally exist) to appear in OpenKH's KH2MapStudio.**

> Both of these remove the need for the hash system entirely.

Every other tool which I've used in the OpenKH suite can recognize these files without using the hash system at all; for example the OpenKH ObjEntryEditor simply allows me to type my own filenames into it and save (without patching first), and from there once it's in my "00objentry.bin", the OpenKH SystemEditor shows those files selectable in it's own drop down lists - which has been great for testing, allowing me to set my custom entries as players and party members for each individual world... 
_However, the final tool I need to recognize these custom entries is the KH2MapStudio, because that is what allows me to spawn them on a map-specific basis_; and unfortunately **KH2MapStudio is not updating to see my custom MDLX entries, which makes them unspawnable** since I can't just type in something it doesn't already recognize... 

This is the whole reason why I am in search of a method to change what KH2MapStudio is looking for from it's own built-in list, so that I can type the names of my new model entries and have them selectable/spawnable within the KH2MapStudio. 

Now that you mention the OpenKH IDX/IMG tool, I realize I've never used it before and have just now opened it for the first time.
I've mounted my patched .iso file with the custom entries already patched into it's .IDX, and then I used the tool to open the .IDX - however, upon looking in the "obj" folder, still my custom entries do not show up... I am absolutely certain this .IDX I'm viewing has my custom entries & hashes within it, because I'm able to re-extract the files using my updated "_Hashpairs.txt" file with CrazyCatz00's Patcher, yet for some reason those entries do not show up when viewing the .IDX with the OpenKH IDX/IMG tool...

> editing the IDX properly with the IDXIMG tool should generate a proper hash for you and insert it into your hash list.

I've looked all over for a button or feature which will allow me to add these entries to the OpenKH .IDX list, but I only see options for extracting each file when right clicking. All of the text boxes with paths and settings for each file are unchangeable... I feel I am definitely missing something in terms of how this tool works... I have now read the documentation on both the GUI and the CLI IDXIMG tools, but I don't see anything about adding files to the .IDX... I thought maybe the answer was to inject a new file into the .iso and it would update automatically; however when attempting that via the CLI tool, I'm greeted with this message: 

"The file obj/P_XEMNAS.mdlx has not been found inside the KH2.IDX, therefore the injection will terminate."

If you could inform me of how to update the .IDX (or list) which OpenKH's KH2MapStudio is specifically using, so that I can then see the new files spawn via KH2MapStudio, it would be a huge help! 

Thanks again for your time, I greatly appreciate it! 🙏

**Comment by Shinobz** [comment, marked as answer]
After months, I figured out that all I had to do to get custom MDLX names working, was to update the 00objentry file within the extracted game folder... I only had it updated in my VFS folder for CrazyCatz00's Dynamic Loader, copying that file to my extraction folder and re-opening it in MapStudio did the trick! No more trying to remember which "P_EX1084561247564915_RAW_RTN_TSURU_51651494651984_NM" is Xemnas... now he's just known as "P_XEMNAS". 

I give my best regards to those trying to solve this in the future... I'd still recommend improving this model-reading system to something a bit easier such as just pulling the model list from the extracted game obj folder... The same way 03system editor does.

*Copied comment from closed issue to make this solution more visible in "Discussions" where it was first asked about, not intended to be spam.
